Improve Left/Right/Inner Join

flcong opened the pull request:
* Fix #190
* For left (right) join, set `pd.NA` to observations not matched by equality or inequality condition from the right (left) table.
* Add new data sets and tests for complex join.

flcong committed:
Improve Left/Right/Inner Join

flcong committed:
Add new datasets in tests

flcong committed:
Fix issue for new datasets

flcong committed:
Add new test for complex join

nils-braun commented:
Thanks @flcong - looks like a lot of work! I might need some time checking the content. Make sure to ping me if I take too long... In the meantime, maybe @rajagurunath can have a first look (although this is also new code for you, sorry!)

flcong commented:
> Thanks @flcong - looks like a lot of work! I might need some time checking the content. Make sure to ping me if I take too long... In the meantime, maybe @rajagurunath can have a first look (although this is also new code for you, sorry!)

No problem! First time contributing to a python package.

rajagurunath commented:
Hi @flcong,

Thanks for fixing this issue and welcome to dask-sql and open source :)

rajagurunath commented on tests/integration/test_join.py:
```diff
@@ -184,3 +184,358 @@ def test_join_literal(c):
+        dfcorrpn[k] = dfcorrpn[k].astype(v)
+
+    # Left Join
+    querypnl = """
```
I have tested this Query using PostgresSQL, it somewhat matches with the correct dataset specified here except for some `dtype` mismatch,  I have one suggestion can Please you add this testing on `test_postgres.py`  as well, where you can make use of `assert_query_gives_same_result` to compare your result directly with Postgres result. let me know if you need any help.

rajagurunath commented on dask_sql/physical/rel/logical/join.py:
```diff
@@ -177,8 +218,91 @@ def merge_single_partitions(lhs_partition, rhs_partition):
+                    how="left",
+                )
+                # Assign pd.NA
+                for v in df.columns[df.columns.str.startswith(other_varpre)]:
```
Just tried running the tests and the same code in jupyter notebook , and got the following error ```TypeError: float() argument must be a string or a number, not 'NAType' ```, if you are not getting this error then it is probably due to environment mismatch error I guess, Or any other guess regarding this Error?
 maybe due to setting `pd.NA` ? what do you think @flcong?

flcong replied:
Which test triggers the error? It seems to be the conversion `float(pd.NA)`?

rajagurunath replied:
For me, first query (left join) was failing in `test_join_lricomplex`
 I will try to provide more context by tomorrow :)

flcong replied:
OK. BTW, I use Python 3.8.10 and Pandas 1.3.2 to run the tests.

rajagurunath replied:
That's a good catch @flcong, The Above Error was raised on  pandas version `1.2.4` and working fine in `pandas==1.3.2`

flcong commented on tests/integration/test_join.py:
```diff
@@ -184,3 +184,358 @@ def test_join_literal(c):
+
+    # Left Join
+    querypnl = """
+        select a.*, b.startdate, b.enddate, b.lk_nullint, b.lk_int, b.lk_str,
```
All right. Let me check. To do the Postgre tests, I need to install docker, right?

rajagurunath replied:
Yes :), if not, a quick hack for testing purpose is you can hardcore any available postgresSQL ip address in the `engine` fixture of test_postgres.py.whichever is easier.

flcong replied:
OK. I've added the tests in `test_postgres.py`. Also I don't know why but I cannot connect to the postgres container using the original `test_postgres.py`. So I exposed port 5432 in `client.containers.run` and change `address` to `"localhost"`. Then, the tests work fine.

flcong committed:
Add new test data in fixture `assert_query_gives_same_result`

* Add new data
* Allow for type conversion
* Allow for specification of whether or not to check dtypes.

flcong committed:
Add test in Postgres

* Add new tests in `test_join.py` in `test_postgres.py` as well.
* Expose port 5432 for postgres container.
* Change address to "localhost"

rajagurunath reviewed (commented):
Maybe Once these changes are done, I request @nils-braun  for further review since I am also no expert in the `dask-sql/physical/rel/logical` part.

nils-braun reviewed (commented):
Ui @flcong, that was probably quite difficult to find out. Nice job. It took me quite some time to dig through the code, so I think it would be very cool if you can add a bit more documentation. Maybe you can use a small example Dataframe in the comments to show, what is actually wrong and how each step helps in solving it (e.g. the three cases you have and then the different merges later which rows they actually add). If this is too much to ask for, I can also take care of it.

I had some smaller comments in the code, but there is also a more general issue I would like to discuss here (and maybe it is just me being too naive): let me recap the problem. When we do a LEFT or RIGHT join with a filter condition, it happens that we get rid of all NAN-rows in the other Dataframe (and therefore the JOIN turns into an inner join) - I did understand this part. In my naive understanding (please correct me) the solution should be rather easy: if in the resulting Dataframe in the end a certain line from the left (or right) Dataframe is missing and we are doing a LEFT (or RIGHT) join, we still need to keep it. Shouldn't here a very simple `join` with the initial left (or right) Dataframe before the merge and on one of your newly created index columns be enough here? Your solution might also work (I did not think through all corner cases, but looks good so far) but it is quite complex and involves multiple merges (which are shuffles in the Dask world). Am I too simple?

flcong commented on tests/integration/test_postgres.py:
```diff
@@ -126,6 +128,92 @@ def test_join(assert_query_gives_same_result):
+        """,
+        ["dates", "startdate", "enddate"],
+        force_dtype="dask",
+        check_dtype=True,
```
I'm not sure how `assert_frame_equal` works to determine if, for example, `2.0` and `2` are identical when `check_dtype=False`. So here I try to make it more explicit that a type cast to the dask dataframe makes the two dataframes identical even if `check_dtype=True`. I guess it conveys more information for developers? (Maybe just my OCD.)

flcong committed:
Roll back to previous join.py and add merge

* Roll back to previous `join.py` but add lines to merge unmatched columns.
* Fix a bug in `merge_single_partitions` where the returned dataframe has an extra column `"common"` that triggers metadata mismatch in the added lines of merging.

flcong committed:
Simplify new tests added

* Since int columns will be converted to float if there are unmatched rows, I use `check_dtype=False` but still use nullable int in the assumed correct table.

flcong commented:
> > Am I too simple?
> > Just to make my point clear: if you replace your changes in `join.py` inside the `filter_condition` check just with
> 
> ```
> logger.debug(f"Additionally applying filter {filter_condition}")
> df = filter_or_scalar(df, filter_condition)
> 
> # make sure we recover any lost rows in case of left, right or outer joins
> if join_type in ["left", "outer"]:
>     df = df.merge(df_lhs_renamed, on=list(df_lhs_renamed.columns), how="right")
> elif join_type in ["right", "outer"]:
>     df = df.merge(df_rhs_renamed, on=list(df_rhs_renamed.columns), how="right")
> 
> df = df.drop(columns=["left_idx", "right_idx"])
> ```
> 
> your tests still succeed

That's brilliant! I didn't think of that. I rolled back `join.py` to the previous version and added the lines you mentioned. 

One caveat is that in this case, if any input column has type int and there are unmatched rows, `np.nan` will be added and the column type will be cast to float, instead of nullable int. Since we don't know whether or not there are unmatched rows without triggering `compute()` (I guess), we have to either leave it as is (users may want to type-cast float into nullable int later) or we cast all int columns to nullable int (which may not be good either). What do you think?

By the way, I also find a bug in the `merge_single_partitions` function in `join.py`. (I don't know how you quote code in the conversation XD.) In the returned dataframe, there is an extra column `"common"` which causes a metadata mismatch for left join when there are only inequality conditions (e.g. my time-series table tests), so I dropped the column.

flcong commented on tests/integration/fixtures.py:
```diff
@@ -86,6 +86,88 @@ def datetime_table():
 
 
+@pytest.fixture
+def user_table_lk():
```
Yeah. I realize that. I simplified the new tests.
